tracer/src/Datadog.Trace/Debugger/Symbols/SymbolExtractor.cs

```diff
@@ -0,0 +1,718 @@
+                closureMethods = ArrayPool<Model.Scope>.Shared.Rent(methods.Count + (nestedTypes.Length * 2));
+                int index = 0;
+
+                for (int i = 0; i < nestedTypes.Length; i++)
+                {
+                    if (!DatadogMetadataReader.IsCompilerGeneratedAttributeDefinedOnType(nestedTypes[i].RowId))
+                    {
+                        continue;
+                    }
+
+                    var nestedType = MetadataReader.GetTypeDefinition(nestedTypes[i]);
+                    var generatedMethods = nestedType.GetMethods();
+                    foreach (var generatedMethodHandle in generatedMethods)
+                    {
+                        if (generatedMethodHandle.IsNil)
+                        {
+                            continue;
+                        }
+
+                        var generatedMethodDef = MetadataReader.GetMethodDefinition(generatedMethodHandle);
+                        PopulateClosureMethod(generatedMethodDef, nestedType);
+                    }
+                }
+
+                foreach (var methodHandle in methods)
+                {
+                    if (methodHandle.IsNil || methodHandle == methodDef.Handle)
+                    {
+                        continue;
+                    }
+
+                    if (!DatadogMetadataReader.IsCompilerGeneratedAttributeDefinedOnMethod(methodHandle.RowId))
+                    {
+                        continue;
+                    }
+
+                    var currentMethod = MetadataReader.GetMethodDefinition(methodHandle);
+                    PopulateClosureMethod(currentMethod, typeDef);
+                }
+
+                if (index == 0)
+                {
+                    return null;
+                }
+
+                var closureScopes = new Model.Scope[index];
+                for (int i = 0; i < index; i++)
+                {
+                    closureScopes[i] = closureMethods[i];
+                }
+
+                return closureScopes;
+
+                void PopulateClosureMethod(MethodDefinition generatedMethod, TypeDefinition ownerType)
+                {
+                    var closureMethodScope = CreateMethodScopeForGeneratedMethod(methodDef, generatedMethod, ownerType);
+                    if (closureMethodScope.HasValue)
+                    {
+                        if (index < closureMethods.Length)
+                        {
+                            closureMethods[index] = closureMethodScope.Value;
+                            index++;
+                        }
+                        else
+                        {
+                            Log.Warning("Not enough space for all closure methods");
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                if (closureMethods != null)
+                {
+                    ArrayPool<Model.Scope>.Shared.Return(closureMethods);
```

Comment:
I don't get it. You rent an array from the array pool, return it (line 532) to the caller and also return it to the array pool?
Don't we have a potential ownership problem in here?

Reply:
Definitely! I tried to catch all of these cases but it turns out I missed it. Thanks

Reply:
You managed to confuse me :) I return a different array, not the same one. (`closureScopes` & `closureMethods`)

Reply:
Oops 😅 , maybe rename the rented array to `rentedArrayClosureScopes` to minimize confusion?

Reply:
Oops 😅 , resolving.